### PR TITLE
APP-2280 Add @viamrobotics/prime-core package

### PR DESCRIPTION
**Pull request by ethanlook**
This adds a new `@viamrobotics/prime-core` package to PRIME. The package is a SvelteKit component library.

## Change log

- `npm create svelte@latest core`
    - Use "component library" option
- Add Badge component
    - Add Badge tests using Vitest
    - Use Badge component in dev environment (not published)
- Update GitHub Workflow to run all checks, builds, and tests
- Update GitHub Workflow to publish NPM package

## Testing

- Added `Badge` tests
- Dry run [npm publish](https://github.com/viamrobotics/prime/actions/runs/5673468740/job/15375029127?pr=282) - please check that the package is structured as expected

## Review requests

Please try to pull down the branch and see if you can use the component in your own local work. Instructions are in the README.

### CODE REVIEW

File: packages/core/.eslintignore
Comment (mcous) [inline packages/core/.eslintignore:1]: I think this was probably the generated SvelteKit default, but it always bugs me that they'll just jam everything in these ignore files without thought, leading to bloated files that don't do anything

File: packages/core/.eslintignore [hunk 1]
```diff
@@ -0,0 +1,14 @@
+.DS_Store
```
Comment (mcous) [inline packages/core/.eslintignore:1]: Not a JS file, ESLint doesn't care

File: packages/core/.eslintignore [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+node_modules
+/build
+/.svelte-kit
+/package
+.env
+.env.*
+!.env.example
```
Comment (mcous) [inline packages/core/.eslintignore:8]: Not files ESLint cares about

File: packages/core/.npmrc [hunk 3]
```diff
@@ -0,0 +1,2 @@
+engine-strict=true
+resolution-mode=highest
```
Comment (mcous) [inline packages/core/.npmrc:2]: Haven't messed with the `resolution-mode` setting, was this something you set? What's reasoning? (Might be worth a comment, especially if we want to set this as a default across other projects)

File: packages/core/.prettierignore
Comment (mcous) [inline packages/core/.prettierignore:1]: Same comments about the `.eslintignore`, with the caveat that that the `pnpm-lock.yaml` ignore is actually important here

File: packages/core/README.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+## Installation
+
+Add PRIME core using your package manager of choice:
+
+```
+pnpm add --save-dev @viamrobotics/prime-core
+```
+
+Install [Tailwind][]. In the `tailwind.config.js`, add the components to the content and include the theme:
```
Comment (mcous) [inline packages/core/README.md:15]: Are we missing a `[tailwind]` anchor href down below?

File: packages/core/package.json [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    },
+    "./prime.css": "./src/prime.css",
+    "./theme.js": "./src/theme.js"
+  },
+  "files": [
+    "dist",
+    "!dist/**/*.test.*",
+    "!dist/**/*.spec.*",
+    "src/prime.css",
+    "src/theme.js"
```
Comment (mcous) [inline packages/core/package.json:30]: If we're generating our output files with source maps, I like to include the whole `src` directory. It increases the download size a little, but doesn't harm the actual bundle size, and can help with debugging

File: packages/core/package.json [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+  ],
+  "peerDependencies": {
+    "svelte": "^4.0.0"
```
Comment (mcous) [inline packages/core/package.json:33]: This is definitely a correct spec, but I like to spec peers with explicit bottom and top values, to make it easier in the future to simply widen the range, rather than mindlessly update the peer and accidentally introduce a breaking change to the package contract.

```suggestion
    "svelte": ">=4.0.0 <5"
```

File: packages/core/package.json [hunk 7]
```diff
@@ -0,0 +1,73 @@
+{
+  "name": "@viamrobotics/prime-core",
+  "version": "0.0.1",
+  "scripts": {
+    "dev": "vite dev",
+    "build": "vite build && npm run package",
```
Comment (mcous) [inline packages/core/package.json:6]:
```suggestion
    "build": "vite build && pnpm run package",
```

File: packages/core/package.json [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    "preview": "vite preview",
+    "package": "svelte-kit sync && svelte-package && publint",
+    "prepublishOnly": "npm run package",
+    "check": "concurrently -g pnpm:check-*",
+    "check-svelte": "svelte-kit sync && svelte-check --tsconfig ./tsconfig.json",
+    "check-lint": "prettier --check . && eslint .",
+    "format": "prettier --write .",
```
Comment (mcous) [inline packages/core/package.json:13]: Are we missing extensions on either of these calls? IIRC (could be wrong) using `.` will only check prettier and eslint defaults, which won't include `.svelte`, `.ts`, etc

File: packages/core/package.json [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    "test": "vitest run",
+    "test:watch": "vitest"
+  },
+  "exports": {
+    ".": {
+      "types": "./dist/index.d.ts",
+      "svelte": "./dist/index.js"
```
Comment (mcous) [inline packages/core/package.json:20]: I assume this is what SvelteKit outputs? A little suprising there isn't a `default` or `import` export, but I guess it makes sense if this only works with Svelte

File: packages/core/package.json [hunk 10]
```diff
@@ -0,0 +1,73 @@
+{
+  "name": "@viamrobotics/prime-core",
+  "version": "0.0.1",
+  "scripts": {
```
Comment (mcous) [inline packages/core/package.json:4]: We'll want a publish config section (there are a few ways to do this, `publishConfig` is my favorite because it's portable) to tell `npm` on the initial publish that this namespaced package should be public, because the npm default for namespaced packages is private

```suggestion
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
```

File: packages/core/vite.config.ts [hunk 11]
```diff
@@ -0,0 +1,11 @@
+import { sveltekit } from '@sveltejs/kit/vite';
+import { defineConfig } from 'vitest/config';
+
+export default defineConfig({
+  plugins: [sveltekit()],
+  test: {
+    include: ['src/**/*.{test,spec}.{js,ts}'],
+    globals: true,
```
Comment (mcous) [inline packages/core/vite.config.ts:8]: Rather than setting `globals: true`, we can set up a `vitest.setup.ts` file to install `@testing-library/jest-dom` and configure the testing-library reset. I softly prefer it to globals mode which feels to me like a silly holdover from pre-module times

**Review by mcous — APPROVED** [review]
Looks fantastic to me! Got some random package management nits and thoughts, but otherwise LGTM!

(Except for that `publishConfig` comment, which is needed or else the first publish will fail)

File: packages/core/.prettierignore
Comment (ethanlook) [reply packages/core/.prettierignore:1]: Why is the `pnpm-lock.yaml` important? It's located in the parent workspace

File: packages/core/package.json (hunk 9 above)
Comment (ethanlook) [reply packages/core/package.json:20]: Yup, this is what SvelteKit outputs. Also a little surprising but it makes sense.

File: packages/core/src/theme.js [hunk 12]
```diff
@@ -0,0 +1,52 @@
+export const theme = {
```
Comment (micheal-parks) [inline packages/core/src/theme.js:1]: Theme is used both in core and legacy correct? are we having legacy import this theme from core? If not I'd probably recommend that since problems created by design drift I think would be noisier worse than problems caused by cross pollinating data between packages in this case

Comment (ethanlook) [reply packages/core/src/theme.js]: Great idea, I just made legacy import from core.

File: .github/workflows/ci.yml [hunk 13]
```diff
@@ -62,21 +67,26 @@ jobs:
...
     steps:
       - uses: actions/checkout@v3
-      - name: 'Install Node.js'
-        uses: actions/setup-node@v3
+      - name: Install 🔧
+        uses: ./.github/actions/setup
         with:
-          node-version: 18
-      - name: Download npm artifacts
+          node-version: 16
```
Comment (mcous) [inline .github/workflows/ci.yml:73]: Node 16 is EOL'ing in a few months; was this change intentional?

File: packages/legacy/.eslintrc.cjs [hunk 14]
```diff
@@ -8,7 +10,10 @@ module.exports = {
   parserOptions: {
     tsConfigRootDir: __dirname,
```
Comment (mcous) [inline packages/legacy/.eslintrc.cjs:11]: Is this casing correct?

```suggestion
    tsconfigRootDir: __dirname,
```

File: packages/legacy/.eslintrc.cjs [hunk 15]
```diff
@@ -8,7 +10,10 @@ module.exports = {
...
     extraFileExtensions: ['.svelte'],
-    project: ['./tsconfig.json', './playground/tsconfig.json'],
+    project: [
+      path.join(__dirname, './tsconfig.json'),
+      path.join(__dirname, './playground/tsconfig.json'),
```
Comment (mcous) [inline packages/legacy/.eslintrc.cjs:15]: These `path.join`'s shouldn't be required; might be related to casing comment above

File: packages/legacy/package.json [hunk 16]
```diff
@@ -67,6 +63,7 @@
     "@typescript-eslint/parser": "^5.59.0",
     "@viamrobotics/eslint-config": "^0.0.4",
     "@viamrobotics/prettier-config": "^0.0.1",
+    "@viamrobotics/prime-core": "workspace:../core",
```
Comment (mcous) [inline packages/legacy/package.json:66]: Oh duh, hadn't considered that we'd added a `workspace` dependency. This makes our usage of `pnpm publish -r` _required_ rather than "nice to have" (to replace `workspace` deps with the actual versions).

Glad you took care of that!

**Review by mcous — APPROVED** [review]
LGTM, might be a little (unharmful) weirdness remaining in the legacy ESLint config because of course there is it's ESLint

File: packages/legacy/package.json (hunk 16 above)
Comment (ethanlook) [reply packages/legacy/package.json:66]: I just added this dependency so we could share the theme across packages, but we didn't do that when you first reviewed. It's much better now anyways!

File: .github/workflows/ci.yml (hunk 13 above)
Comment (ethanlook) [reply .github/workflows/ci.yml:73]: Nope - copy pasta

File: packages/legacy/.eslintrc.cjs (hunk 14 above)
Comment (ethanlook) [reply packages/legacy/.eslintrc.cjs:11]: It was not! Great catch.

File: packages/legacy/.eslintrc.cjs (hunk 15 above)
Comment (ethanlook) [reply packages/legacy/.eslintrc.cjs:15]: Unfortunately the casing did not affect the need for the `path.join`s. Leaving as-is.
